modules/contact-form/admin.php

```diff
@@ -148,17 +146,43 @@ function grunion_add_empty_spam_button() {
 		return;
 	}
 
-	// Get HTML for the button
-	$button_html  = wp_nonce_field( 'bulk-destroy', '_destroy_nonce', true, false );
-	$button_html .= get_submit_button( __( 'Empty Spam', 'jetpack' ), 'apply', 'delete_all', false );
+	$feedbacks_count = wp_count_posts( 'feedback' );
+	$nonce           = wp_create_nonce( 'jetpack_delete_spam_feedbacks' );
+	$success_url     = remove_query_arg( array( 'jetpack_empty_feedback_spam_error', 'post_status' ) ); // Go to the "All Feedback" page.
+	$failure_url     = add_query_arg( 'jetpack_empty_feedback_spam_error', '1' ); // Refresh the current page and show an error.
+	$spam_count      = $feedbacks_count->spam;
 
-	// Add the button next to the filter button via js
 	?>
-		<script type="text/javascript">
-			jQuery(document).ready(function($) {
-				$('#posts-filter #post-query-submit').after('<?php echo $button_html; ?>' );
-			})
-		</script>
+	<script type="text/javascript">
+		jQuery( function ( $ ) {
+			var jetpack_empty_spam_feedbacks_button_container = $( '<div/>' )
+				.addClass( 'jetpack-empty-spam-container' );
+
+			var jetpack_empty_spam_feedbacks_button = $( '<a />' )
+				.addClass( 'button-secondary' )
+				.addClass( 'jetpack-empty-spam' )
+				.attr( 'href', '#' )
+				<?php /* translators: The placeholder is for showing how much of the process has completed, as a percent. e.g., "Emptying Spam (40%)" */ ?>
+				.attr( 'data-progress-label', <?php echo wp_json_encode( __( 'Emptying Spam (%1$s%)', 'jetpack' ) ); ?> )
+				.attr( 'data-success-url', <?php echo wp_json_encode( $success_url ); ?> )
+				.attr( 'data-failure-url', <?php echo wp_json_encode( $failure_url ); ?> )
+				.attr( 'data-spam-feedbacks-count', <?php echo wp_json_encode( $spam_count ); ?> )
+				.attr( 'data-nonce', <?php echo wp_json_encode( $nonce ); ?> )
+				.text( <?php echo wp_json_encode( __( 'Empty Spam', 'jetpack' ) ); ?> )
+				;
+
+			jetpack_empty_spam_feedbacks_button_container.append( jetpack_empty_spam_feedbacks_button );
+
+			var jetpack_empty_spam_feedbacks_spinner = $( '<span />' )
+				.addClass( 'jetpack-empty-spam-spinner' )
+				;
+
+			jetpack_empty_spam_feedbacks_button_container.append( jetpack_empty_spam_feedbacks_spinner );
+
+			// Add the button both above and below the list of spam feedbacks.
+			$( '.tablenav.top .actions, .tablenav.bottom .actions' ).not( '.bulkactions' ).append( jetpack_empty_spam_feedbacks_button_container );
+		} );
+	</script>
```

Comment:
Part of me wonders if it wouldn't be cleaner to build the button in `grunion-admin.js` right away, and rely on `wp_localize_script` for all the variables. What do you think?

Reply:
Good call; I've done so in 78123282f.